src/core/controllers/map-controller.js

```diff
@@ -0,0 +1,286 @@
+
+  let startZoom = startViewState.props.zoom;
+
+  if (!Number.isFinite(startZoom)) {
+    startZoom = viewState.props.zoom;
+  }
+
+  // take the start lnglat and put it where the mouse is down.
+  const newZoom = startZoom + Math.log2(scale);
+  const zoomedViewport = new WebMercatorViewport(
+    Object.assign({}, viewState.props, {
+      zoom: newZoom
+    })
+  );
+  const [longitude, latitude] = zoomedViewport.getLocationAtPoint({lngLat: startZoomLngLat, pos});
+
+  return viewState.getUpdatedState({
+    zoom: newZoom, // Rely on constraints being applied
+    longitude,
+    latitude
+  });
+}
+
+function zoomIn(viewState) {
+  return viewState.getUpdatedState({
+    zoom: viewState.props.zoom + 0.2
+  });
+}
+
+function zoomOut(viewState) {
+  return viewState.getUpdatedState({
+    zoom: viewState.props.zoom - 0.2
+  });
+}
+
+function moveLeft(viewState) {
+  return viewState.getUpdatedState({
+    bearing: viewState.props.bearing - 3
+  });
+}
+
+function moveRight(viewState) {
+  return viewState.getUpdatedState({
+    bearing: viewState.props.bearing + 3
+  });
+}
+
+function moveForward(viewState) {
+  return viewState.getUpdatedState({
+    pitch: viewState.props.pitch + 3
+  });
+}
+
+function moveBackward(viewState) {
+  return viewState.getUpdatedState({
+    pitch: viewState.props.pitch - 3
+  });
+}
+
+export const EVENT_REDUCERS = {
+  // panStart,
+  pan,
+  // panEnd,
+  // rotateStart,
+  rotate,
+  // rotateEnd,
+  // zoomStart,
+  zoom: zoomTo,
+  // zoomEnd,
+  zoomIn,
+  zoomOut,
+  moveLeft,
+  moveRight,
+  moveForward,
+  moveBackward
+};
+
+// A class that handles events and updates mercator style viewport parameters
+export default class MapController extends Controller {
+  constructor(options) {
+    // Register map specific "ViewState + event -> ViewState" reducers
+    super(options, {
+      defaultReducers: EVENT_REDUCERS,
+      defaultConstraints: MAPBOX_LIMITS
+    });
+  }
+
+  getViewportProps(viewState) {
+    return this._applyConstraints(viewState, this.limits);
+  }
+
+  _applyConstraints(viewState, constraints) {
+    const {width, height} = this.getSize();
+
+    const {
+      longitude,
+      latitude,
+      zoom,
+      pitch = 0,
+      bearing = 0,
+      altitude = 1.5,
+
+      maxZoom,
+      minZoom,
+      maxPitch,
+      minPitch,
+      maxLatitude,
+      minLatitude
+    } = Object.assign({}, constraints, viewState.props);
+
+    const props = {
+      width,
+      height,
+      longitude,
+      latitude,
+      zoom,
+      pitch,
+      bearing,
+      altitude,
+      maxZoom,
+      minZoom,
+      maxPitch,
+      minPitch,
+      maxLatitude,
+      minLatitude
+    };
+
+    props.longitude = normalizeDegrees(props.longitude);
+    props.bearing = normalizeDegrees(props.bearing);
+    props.zoom = clamp(zoom, minZoom, maxZoom);
+    props.pitch = clamp(pitch, minPitch, maxPitch);
+
+    return normalizeViewportProps(props);
```

Comment:
`normalizeViewportProps` only returns `{width, height, longitude, latitude, zoom, pitch, bearing}`. Other props will be lost.